[SPARK-53573][SQL] Use Pre-processor for generalized parameter marker handling

srielau opened the pull request:
## What changes were proposed in this pull request?

This PR introduces a comprehensive parameter substitution system that significantly expands parameter marker support and implements major performance optimizations. The current implementation of parameter markers (`:param` and `?`) is restricted to expressions and queries. This PR enables parameter markers to work universally across all SQL constructs.

### Key Changes

**Enhanced Parameter Substitution Architecture:**
- Added `ParameterHandler` class with centralized, optimized parameter substitution logic
- Integrated parameter substitution directly into `SparkSqlParser` with parser-aware error context
- Implemented `ParameterSubstitutionContext` with thread-local substitution information
- Added sophisticated position mapping for accurate error reporting in original SQL text

**EXECUTE IMMEDIATE Integration & Error Context:**
- Enhanced `ResolveExecuteImmediate` with comprehensive parameter validation and local variable isolation
- Added `HybridParameterContext` supporting both named and positional parameters with strict validation
- Implemented parser-aware error context that correctly maps errors back to original SQL before parameter substitution
- Added "firewall" mechanism preventing `EXECUTE IMMEDIATE` from accessing outer SQL scripting variables

**Legacy Mode Support:**
- Maintained full backward compatibility with `spark.sql.legacy.parameterSubstitution.constantsOnly`
- In legacy mode: parameter substitution disabled, analyzer-based binding preserved
- In new mode: full text-based parameter substitution with enhanced validation

## Why are the changes needed?

**Functional Limitations:**
1. **Limited scope**: Parameter markers only worked in expressions/queries, not DDL/utility statements
2. **Analyzer dependency**: Literals bypass analyzer, making parameter expansion impossible

**The enhanced architecture solves these by:**
- Universal parameter support through pre-parsing substitution
- Parser-aware error context maintaining accurate error positions
- High-performance algorithms optimized for large SQL texts
- Comprehensive validation preventing common parameter usage errors

## Does this PR introduce _any_ user-facing changes?

**Yes - Significantly Expanded Functionality:**

**New Parameter Support:**
- DDL statements: `CREATE VIEW v(c1) AS (SELECT :p)`
- Utility statements: `SHOW TABLES FROM schema LIKE :pattern`
- Type definitions: `DECIMAL(?, ?)`
- Table properties: `TBLPROPERTIES (':key' = ':value')`
- All SQL constructs where literals are valid

**Full Backward Compatibility:**
- All existing parameter usage continues working unchanged
- Legacy mode behavior completely preserved
- No breaking changes to existing APIs

## How was this patch tested?

**Comprehensive Unit Testing:**
- Enhanced `ParametersSuite` with 50+ test cases covering all parameter scenarios
- Added `ParameterSubstitutionSuite` for algorithm-specific testing
- `SqlScriptingExecutionSuite` tests for EXECUTE IMMEDIATE isolation
- Position mapping accuracy tests ensuring correct error reporting

**Integration Testing:**
- Cross-validated legacy vs modern mode behavior consistency
- EXECUTE IMMEDIATE parameter validation across all scenarios
- Error context accuracy for nested parameter substitution
- Thread safety validation for concurrent parameter usage

**Manual Verification:**
- DDL parameter substitution: `CREATE TABLE :name (id INT)`
- Complex parameter scenarios: `EXECUTE IMMEDIATE 'SELECT typeof(:p), :p' USING 5::INT AS p`
- Error position accuracy in multi-line SQL with parameters
- Performance benchmarking on large SQL scripts

**Example Enhanced Capabilities:**
```scala
// Universal parameter support
spark.sql("CREATE TABLE :table_name (id :type)", 
  Map("table_name" -> "users", "type" -> "BIGINT"))

// Enhanced validation with accurate error context
spark.sql("EXECUTE IMMEDIATE 'SELECT :p' USING 1, 2 AS p") 
// Fails with ALL_PARAMETERS_MUST_BE_NAMED, showing exact position

// Performance: handles large SQL with many parameters efficiently
spark.sql(largeSqlWithManyParameters, parameterMap) // Now O(k) instead of O(n²)
```

**Quality Assurance:**
- All scalastyle checks pass
- No compilation warnings or deprecation issues
- Thread safety verified through concurrent testing
- Memory usage profiling confirms optimization effectiveness

cloud-fan commented on sql/api/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4:
```diff
@@ -40,6 +40,12 @@ options { tokenVocab = SqlBaseLexer; }
+   * When true, parameter markers are allowed everywhere a literal is supported.
+   */
+  public boolean parameter_substitution_enabled = true;
 }
```
This PR does not touch the CREATE TABLE parser rule, how do we support parameter markers there? Or it's already supported before this PR?

cloud-fan commented on sql/api/src/main/scala/org/apache/spark/sql/catalyst/parser/DataTypeAstBuilder.scala:
```diff
@@ -138,7 +165,7 @@ class DataTypeAstBuilder extends SqlBaseParserBaseVisitor[AnyRef] {
     } else {
       val badType = typeCtx.unsupportedType.getText
-      val params = typeCtx.INTEGER_VALUE().asScala.toList
+      val params = typeCtx.integerValue().asScala.map(_.getText).toList
```
So the assumption here is the parameter markers should have been replaces and we can safely get the original INTEGER_VALUE text?

cloud-fan commented on sql/api/src/main/scala/org/apache/spark/sql/catalyst/parser/DataTypeAstBuilder.scala:
```diff
@@ -258,9 +285,15 @@ class DataTypeAstBuilder extends SqlBaseParserBaseVisitor[AnyRef] {
    */
   override def visitCommentSpec(ctx: CommentSpecContext): String = withOrigin(ctx) {
-    string(visitStringLit(ctx.stringLit))
+    string(visit(ctx.stringLit).asInstanceOf[Token])
```
why do we need this change given we've added a new `visitStringLit` below?

cloud-fan commented on sql/api/src/main/scala/org/apache/spark/sql/catalyst/parser/SubstituteParmsAstBuilder.scala:
```diff
@@ -0,0 +1,255 @@
+      val stopIndex = ctx.getStop.getStopIndex + 1
+      namedParamLocations.getOrElseUpdate(
+        paramName,
+        scala.collection.mutable.ListBuffer[ParameterLocation]()) +=
+        ParameterLocation(startIndex, stopIndex)
```
```suggestion
        scala.collection.mutable.ListBuffer[ParameterLocation]()
      ) += ParameterLocation(startIndex, stopIndex)
```

cloud-fan commented on sql/api/src/main/scala/org/apache/spark/sql/catalyst/parser/SubstituteParmsAstBuilder.scala:
```diff
@@ -0,0 +1,255 @@
+  /**
+   * Create a named parameter which represents a literal with a non-bound value and unknown type.
+   */
+  override def visitNamedParameterLiteral(ctx: NamedParameterLiteralContext): String =
```
does the return value matter?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParameterContext.scala:
```diff
@@ -0,0 +1,87 @@
+ * from SparkSession to the parser without changing method signatures.
+ */
+object ThreadLocalParameterContext {
+  private val context = new ThreadLocal[Option[ParameterContext]]()
```
can we use `LexicalThreadLocal`?

srielau commented on sql/api/src/main/scala/org/apache/spark/sql/catalyst/parser/SubstituteParmsAstBuilder.scala:
```diff
@@ -0,0 +1,255 @@
+   * Handle namedParameterValue context for parameter markers in string literal contexts. This
+   * handles the namedParameterMarker case added to the stringLit grammar rule.
+   */
+  override def visitNamedParameterValue(ctx: NamedParameterValueContext): String =
```
Yes they are, but we can't share the visits due to the way antlr works. Given how short they are, I'd prefer to keep the duplication and later we can remove the old one with teh old way of doping parameters.

srielau commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParameterHandler.scala:
```diff
@@ -0,0 +1,350 @@
+   */
+  def determineSubstitutionRule(sqlText: String): SubstitutionRule = {
+    val trimmed = sqlText.trim
+    if (trimmed.length >= 5 && trimmed.regionMatches(true, 0, "BEGIN", 0, 5)) {
```
This entire rule determination stuff is obsolete. It originated from the time when I tried to parse expressions etc selectively...
I'm removing this API.

srielau commented on sql/core/src/main/scala/org/apache/spark/sql/classic/SparkSession.scala:
```diff
@@ -448,16 +448,26 @@ class SparkSession private(
+          // Set parameter context for parsing
+          val paramContext = PositionalParameterContext(args.map(lit(_).expr).toSeq)
+          ThreadLocalParameterContext.withContext(paramContext) {
+            val parsed = sessionState.sqlParser.parsePlan(sqlText)
```
The flow is:

1. SparkSession sets up parameter context in ThreadLocal
2. SparkSession calls sqlParser.parsePlan(sqlText)
3. SparkSqlParser.parse automatically detects the ThreadLocal parameter context
4. SparkSqlParser calls the pre-parser (ParameterHandler.substituteParameters)
5. Pre-parser returns substituted SQL text (e.g., "SELECT ?" → "SELECT 42")
6. SparkSqlParser continues with normal ANTLR parsing on the substituted text

srielau committed:
Implement parameter substitution pre-parser for EXECUTE IMMEDIATE

This commit implements a comprehensive parameter substitution system for Spark SQL
that provides detailed error context for EXECUTE IMMEDIATE statements, similar to
how views handle errors.

Key features:
- Pre-parser approach with position mapping for accurate error reporting
- Thread-local parameter context for parser-aware error handling
- Support for both named and positional parameters across all SQL APIs
- Optimized position mapping algorithm (O(n²) → O(k) where k = substitutions)
- Comprehensive test coverage including edge cases and error scenarios
- Backward compatibility with legacy parameter substitution mode

The implementation includes:
- ParameterHandler for unified parameter processing
- PositionMapper for efficient error position translation
- LiteralToSqlConverter for type-safe SQL generation
- Integration with SparkSqlParser, SparkSession, and EXECUTE IMMEDIATE
- Enhanced error messages showing both outer and inner query context

This addresses the user request for detailed error context in EXECUTE IMMEDIATE
statements while maintaining performance and compatibility.

srielau committed:
Fix ProcedurSuite and ParameterPositionMappingSuite

miland-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:
```diff
@@ -55,6 +55,9 @@ object SQLExecution extends Logging {
 
   private val testing = sys.props.contains(IS_TESTING.key)
 
+
+
+
```
nit: empty lines

miland-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:
```diff
@@ -49,10 +52,110 @@ class SparkSqlParser extends AbstractSqlParser {
+    }
+  }
+
+  // Override parsing methods that should NOT use parameter substitution
+  // These methods parse identifiers and data types where parameters don't make sense
```
Most comments in Spark follow the convention of using complete sentences and ending them with a punctuation. We should try to maintain that consistency here. Relevant discussion: https://github.com/apache/spark/pull/48950#discussion_r1857421840

srielau commented on sql/core/src/test/scala/org/apache/spark/sql/ParametersSuite.scala:
```diff
@@ -780,9 +767,9 @@ class ParametersSuite extends QueryTest with SharedSparkSession {
 
     checkAnswer(spark.sql(query(":cte"), args = Map("cte" -> "t1")),
-      Seq(Row(1), Row(2), Row(3), Row(4), Row(5)))
+      Array(Row(1), Row(2), Row(3), Row(4), Row(5)))
```
Probably not.. Reverting

srielau committed:
Address comments by Milan

srielau committed:
Remove unused threadlocal

srielau committed:
Remove thread local

srielau committed:
More cleanup

srielau committed:
fix origin comparison

srielau committed:
Fix commnets

srielau committed:
cleanup

srielau committed:
Lint

srielau committed:
Fix testcase

srielau committed:
Remove another thread local to address comments by wenchen

srielau committed:
more review comments

srielau committed:
Remove callback

srielau committed:
Review comments

srielau committed:
fixes

srielau committed:
Add PositionMapper

srielau committed:
fix failing testcases

srielau committed:
fix testcases

dtenedor reviewed (commented):
Did another full review -- this approach is looking a lot simpler without the callbacks :)

srielau committed:
Address comments by Daniel

dtenedor reviewed (approved):
Thanks for all the work on this, it will make our SQL language more usable.

srielau committed:
Fix query plan suite

srielau committed:
More review xomments

srielau committed:
Address review commente by wenchen

srielau committed:
Remove extra parse step for parameter detection.

srielau committed:
Add asserts

srielau committed:
Simplify origin management

srielau committed:
Fix execute immediate

srielau committed:
Fix positionmapper

srielau committed:
Fix queryplansuite

srielau committed:
more fixes

srielau committed:
Fix position mapper - again

srielau committed:
Merge branch 'master' into preparser-squashed

srielau committed:
Consolidate testcases

srielau commented on sql/core/src/main/scala/org/apache/spark/sql/classic/SparkSession.scala:
```diff
@@ -448,16 +448,24 @@ class SparkSession private(
+        val parsedPlan = if (args.nonEmpty) {
+          // Use parameter context directly for parsing
+          val paramContext = PositionalParameterContext(args.map(lit(_).expr).toSeq)
+          val parsed = sessionState.sqlParser.parsePlanWithParameters(sqlText, paramContext)
```
I don't think so. Tehre is stuiff in parsePLanWithParameters which we need in legacy mode.

srielau committed:
More comments by wenchen

srielau commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:
```diff
@@ -50,9 +54,152 @@ class SparkSqlParser extends AbstractSqlParser {
+  private def parseInternal[T](
+      command: String,
+      parameterContext: Option[ParameterContext],
+      isTopLevel: Boolean)
```
outdated

srielau committed:
More comments

srielau committed:
Cleanup

cloud-fan commented on sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkParserUtils.scala:
```diff
@@ -217,14 +225,36 @@ trait SparkParserUtils {
+
+    // Don't map positions yet - store them as-is along with the mapper
+    // Position mapping will be applied when creating query context for errors
+    val (startIndex, stopIndex, text, mapper) =
```
`startIndex` and `stopIndex` are the same between these two branches, we don't need to put them here, but just `val (text, mapper) =`

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParameterHandler.scala:
```diff
@@ -0,0 +1,240 @@
+
+    // Store the position mapper to enable accurate error position reporting in the original SQL.
+    val currentOrigin = CurrentOrigin.get
+    CurrentOrigin.set(currentOrigin.copy(positionMapper = Some(positionMapper)))
```
This is a side effect that updates a global variable. I'm a bit worried about it as we never restore it.

cloud-fan replied:
I think it's safer to return the position mapper to the caller, and let caller to set up the `CurrentOrigin` before invoking the main parser, and restore it afterwards.

srielau committed:
More wenchen comments

srielau committed:
Fix testcases

srielau committed:
Refactor preameter substitution

srielau committed:
fix connect

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/classic/SparkSession.scala:
```diff
@@ -430,6 +431,77 @@ class SparkSession private(
+      // For non-literal allowed expressions, evaluate them to literals
+      val literal = resolvedExpr match {
+        case lit: Literal => lit
+        case foldable if foldable.foldable =>
```
I don't think this is needed, optimizer should have done it already.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/classic/SparkSession.scala:
```diff
@@ -488,15 +573,36 @@ class SparkSession private(
+
+        Dataset.ofRows(self, plan, tracker)
+      } else {
+        // Reset position mapper to prevent contamination from previous parameterized sql() calls
```
I think this is no longer needed, as we reset it right after the parsing: https://github.com/apache/spark/pull/52334/files#diff-77a9aad2da3dc60210a2c4d2f3165d5f1d0acd54ca4811072a053225170ed748R153

srielau committed:
More comments

srielau committed:
Remove origin resets

srielau commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/QueryPlanSuite.scala:
```diff
@@ -37,7 +37,11 @@ class QueryPlanSuite extends SparkFunSuite {
     val mappedQuery = query mapExpressions {
-      case _: Expression => Literal(1)
+      case _: Expression =>
+        // Reset CurrentOrigin to prevent contamination from previous operations
```
Fixed
